src/cargo/core/compiler/job_queue.rs

```diff
@@ -395,7 +395,8 @@ impl<'cfg> JobQueue<'cfg> {
             .filter(|dep| {
                 // Binaries aren't actually needed to *compile* tests, just to run
                 // them, so we don't include this dependency edge in the job graph.
-                !dep.unit.target.is_test() && !dep.unit.target.is_bin()
+                (!dep.unit.target.is_test() && !dep.unit.target.is_bin())
+                    || dep.unit.artifact.is_true()
```

Comment:
Can you say why this was added?

FWIW, I think there is a harmless mistake in the original code.  `dep.unit.target.is_test()` is never true.  I think it was supposed to be `unit.target.is_test()`.  When this code was introduced, the variable `unit` was shadowed causing some confusion.

Reply:
Oh, I think I understand. This is just ensuring that artifact binaries *are* included.  Nevermind my comment.